### PR TITLE
fix: block 7z symlinks before extraction

### PR DESCRIPTION
**Pull request by mldangelo-oai**
## Summary
- reject 7z symlink and junction metadata before any probe or full extraction
- exclude every member in duplicate/canonical-name collision groups because py7zr name targets select all matching physical records
- index archive metadata once across supported py7zr APIs and fail closed on ambiguous or unavailable records
- preserve post-extraction link defense while deduplicating repeated preflight findings

## Critical review
- reproduced a same-name regular-plus-symlink bypass where py7zr extraction selected both records after only the first was preflighted
- reproduced Windows junction metadata being classified as regular
- found duplicate critical reports for named security links checked in both probe and extraction phases
- removed the production special case for `unittest.mock`; tests now declare regular-member metadata explicitly

## Validation
- focused SevenZip hardening class: `40 passed, 35 skipped`
- complete associated scanner module: `65 passed, 54 skipped`
- optional real-py7zr tests skipped locally because py7zr is not installed in this environment
- Ruff format/lint clean
- mypy clean on the changed scanner and test files
- `git diff --check` clean

The branch is synced with current `main`.

### CODE REVIEW

**Comment by mldangelo-oai** [comment]
@codex review

**Comment by mldangelo-oai** [comment]
@codex review

File: modelaudit/scanners/sevenzip_scanner.py [hunk 1]
```diff
@@ -884,6 +891,77 @@ def _read_member_probe_prefix(self, archive: Any, file_name: str, limit: int) ->
...
+    @staticmethod
+    def _preflight_member_symlink_state(member_info: Any | None) -> str:
+        if member_info is None:
+            return "unknown"
+        is_symlink = getattr(member_info, "is_symlink", None)
+        if isinstance(is_symlink, bool):
+            return "symlink" if is_symlink else "regular"
+        if type(is_symlink).__module__.startswith("unittest.mock") and "is_symlink" not in vars(member_info):
+            return "regular"
+        return "unknown"
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/sevenzip_scanner.py:965]: Fixed in aee3deb5. The metadata lookup now prefers `archive.files` entries, whose internal `ArchiveFile.is_symlink` property exists on supported py7zr 0.20.1, before newer public `FileInfo` APIs. Member-size lookup uses the same compatibility helper, and a two-sided legacy regression covers both regular and symlink entries.

File: modelaudit/scanners/sevenzip_scanner.py [hunk 2]
```diff
@@ -884,6 +891,78 @@ def _read_member_probe_prefix(self, archive: Any, file_name: str, limit: int) ->
         probe.seek(0)
         return probe.read(limit)
 
+    def _get_archive_member_info(self, archive: Any, file_name: str) -> Any | None:
+        """Return metadata for a member without extracting it, across py7zr API variants."""
+        files = getattr(archive, "files", None)
+        if isinstance(files, list):
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply modelaudit/scanners/sevenzip_scanner.py]: Fixed in ccbb0b91. `_get_archive_member_info()` now iterates any available `archive.files` container under exception suppression instead of requiring a plain `list`, preserving the public metadata fallback if the internal iterable cannot be consumed. The legacy regression now uses a custom non-list `ArchiveFileList`-style iterable and covers both regular and symlink states.

**Comment by mldangelo-oai** [comment]
Critical review reproduced a pre-extraction bypass: py7zr name targets select every same-name physical entry, so a regular `dup.pkl` followed by a symlink `dup.pkl` was extracted after only the first record was preflighted. The pushed head excludes all ambiguous duplicate/canonical-path groups, blocks junction metadata, indexes member metadata once, and deduplicates probe/extraction link reports.

Validation against the isolated worktree: focused hardening `40 passed, 35 skipped`; complete SevenZip module `65 passed, 54 skipped`; Ruff, mypy, and diff checks clean. Optional real-py7zr cases are skipped locally because that extra is unavailable here. Moving on while CI runs.

**Comment by mldangelo-oai** [comment]
Final QA: installed the current sevenzip extra and ran the complete associated scanner module: 109 passed, 10 skipped. Ruff check, Ruff format check, mypy, and git diff --check are clean. I also tested against the declared minimum py7zr 0.20.1: this branch materially improves compatibility versus main (90 passing vs 62), and the remaining old-version failures are pre-existing writer-factory probe API gaps rather than regressions in this link preflight. Upstream source review confirms both py7zr 0.20.1 and current expose symlink/junction state through the internal ArchiveFile records used here. All review threads are resolved.
